### ISSUE TITLE
Implement a decorator to feed start requests into the scheduler

### ISSUE THREAD

**Issue by Gallaecio** [scrapy/scrapy#5282]
@kmike described the issue best at https://github.com/scrapy/scrapy/issues/456#issuecomment-247622571:
> […] start_requests are not exhausted:
> 
> 1. Scrapy calls start_requests and gets enough requests to fill downloader
> 2. When new requests are scheduled (e.g. from responses) then scrapy pauses getting more requests from start_requests.
> 
> This is inconvenient if you e.g. have 100K websites to crawl and want to crawl their front pages (requests issued in start_requests), and follow some links on them (extracted in parse method). In this case even if you set a very high priority for requests sent in start_requests, they won't be processed. One have to use private APIs like `self.crawler.engine.schedule(request)` to make sure all 100K initial requests are scheduled.
> 
> It seems this is done in order to
> 
> 1. support start_requests with infinite generators;
> 2. send less requests to scheduler at the same time,
> 
> but I think this is implicit and confusing (and not really documented) […].

I think we could address this by providing a documentation topic about start requests, the scheduler, and other aspects related to request consumption order.

The documentation should describe how to achieve alternatives ways of request consumption, so along with the new documentation topic I would provide the following decorator, which addresses the issue in a way that that I prefer over alternatives I have seen proposed before, because instead of letting the Scrapy engine decide when to consume start requests, it gives that power to the scheduler, which I think is the ideal component to determine request order.

I already have an implementation for the decorator that I think can be used as is:

```py
def feed_scheduler(start_requests):
    """Use as a decorator for start_requests to feed its requests into the
    scheduler instead of yielding them directly."""
    def wrapper(spider):
        scheduler = spider.crawler.engine.slot.scheduler
        for request in start_requests(spider):
            scheduler.enqueue_request(request)
        request = scheduler.next_request()
        while request:
            yield request
            request = scheduler.next_request()
    return wrapper
```

I’m not creating a pull request for the decorator because I think we should only merge it along with the documentation I mentioned, which will take me some time to write properly. But I think this decorator can be useful already for people dealing with this issue, so I’m sharing it here already for easy copy-pasting into existing projects.

**Comment by kmike** [comment]
Another idea which we had is to change the behavior for `async def start_requests`. The current behavior almost looks like a bug.

**Comment by t-i-a-n-a** [comment]
@Gallaecio Is the decorator intended to be pasted into the spider class definition, where references to ```spider``` change to ```self```? Then```start_requests``` is passed to the decorator as ```self.start_requests```, I'm assuming by making it a spider attribute? I'm using scrapy version 2.6.3.

```python
def feed_scheduler(start_requests):
    """Use as a decorator for start_requests to feed its requests into the
    scheduler instead of yielding them directly."""
    def wrapper(self):
        scheduler = self.crawler.engine.slot.scheduler
        for request in start_requests(self):
            scheduler.enqueue_request(request)
        request = scheduler.next_request()
        while request:
            yield request
            request = scheduler.next_request()
    return wrapper
```


```python
def __init__(self, *args, **kwargs):
    super(insert_spider_class_name_here, self).__init__(*args, **kwargs)
    self.start_requests = self.feed_scheduler(self.start_requests)
```

Regardless, how would the generator work if the crawl function (in crawler.py) calls start_requests with ```iter(self.spider.start_requests())``` before ```self.engine.open_spider(self.spider, start_requests)``` (in crawler.py) is able to define ```self.crawler.engine.slot.scheduler```? I didn't see another start_requests function call, so my assumption is that the one in crawler.py is the only one.

**Comment by Gallaecio** [comment]
> Is the decorator intended to be pasted into the spider class definition, where references to spider change to self?

The intended usage from what I recall is:

```python
class MySpider:
    # …
    @feed_scheduler
    def start_requests(self):
        # …
```

> how would the decorator work if the crawl function (in crawler.py) calls start_requests with `iter(self.spider.start_requests())` before `self.engine.open_spider(self.spider, start_requests)` (in crawler.py) is able to define `self.crawler.engine.slot.scheduler`?

I have not checked for sure, but I successfully used the decorator in a project, so I think it works.

I believe it works because `iter()` does not cause the code to be called, it is only once `next()` is used that the code is called, and by that time the scheduler should have been initialized.

**Comment by AdrianAtZyte** [comment]
Obsolete since the start() rework in 2.13: start requests all go through the scheduler now, so priority works, and start request order (https://docs.scrapy.org/en/latest/topics/scheduler.html#start-request-order) plus delaying start request iteration (https://docs.scrapy.org/en/latest/topics/spiders.html#delaying-start-request-iteration) cover the documentation side. The decorator is no longer needed.
